rust/agents/relayer/src/msg/gelato_submitter.rs

```diff
@@ -1,98 +1,231 @@
-    /// rate-limiter against the relevant Gelato HTTP endpoint or
-    /// something, or a max-inflight-cap on Gelato messages from
-    /// relayers, enforced here. But probably not until that proves to
-    /// be necessary.
     async fn work_loop(&mut self) -> Result<()> {
         loop {
             self.tick().await?;
-            tokio::task::yield_now().await;
+            sleep(Duration::from_millis(1000)).await;
         }
     }
 
-    /// Extracted from main loop to enable testing submitter state
-    /// after each tick, e.g. in response to a change in environment
-    /// conditions like values in InterchainGasPaymaster.
-    async fn tick(&self) -> Result<()> {
-        // TODO(webbhorn): Pull all available messages out of self.rx
-        // and check if enough gas to process them. If not, put on
-        // wait queue. If there is enough, spawn root task and run
-        // the fwd req op.
+    async fn tick(&mut self) -> Result<()> {
+        // Pull any messages sent by processor over channel.
+        loop {
+            match self.messages.try_recv() {
+                Ok(msg) => {
+                    let op = ForwardRequestOp {
+                        args: self.make_forward_request_args(msg)?,
+                        opts: ForwardRequestOptions::default(),
+                        signer: self.signer.clone(),
+                        http: self.http.clone(),
+                    };
+                    tokio::spawn(async move {
+                        op.run()
+                            .await
+                            .expect("failed unimplemented forward request submit op");
+                    });
+                }
+                Err(TryRecvError::Empty) => {
+                    break;
+                }
+                Err(_) => {
+                    bail!("Disconnected receive channel or fatal err");
+                }
+            }
+        }
+        Ok(())
+    }
 
-        // TODO(webbhorn): Scan pending queue for any newly-eligible
-        // ops and if encountered, spawn them in a root task.
-        // Remove them from pending queue if so.
+    fn make_forward_request_args(&self, msg: SubmitMessageArgs) -> Result<ForwardRequestArgs> {
+        let ivm_base_contract = BaseContract::from(ivm_abi.clone());
+        let call_data = ivm_base_contract.encode(
+            "process",
+            [
+                Token::Address(self.inbox_address),
+                Token::FixedBytes(msg.checkpoint.checkpoint.root.to_fixed_bytes().into()),
+                Token::Uint(msg.checkpoint.checkpoint.index.into()),
+                Token::Array(
+                    msg.checkpoint
+                        .signatures
+                        .iter()
+                        .map(|s| Token::Bytes(s.to_vec()))
+                        .collect(),
+                ),
+                Token::Bytes(msg.committed_message.message.to_vec()),
+                Token::FixedArray(
+                    (0..32)
```

Comment:
nit:
```rust
Token::FixedArray(
  msg.proof.path[0..32].iter().map(|i| Token::FixedBytes(i.to_vec())).collect()
)
```

Basically the same, but I prefer starting with the thing you are iterating when possible rather than an index into it. In this case it isn't hard so this would be my suggestion.

Reply:
Sure, seems perhaps more readable to begin with the array name, I've changed it (but changed the closure's argument name from i to e since it's not capture an index into the array but rather a reference to the element at each index, iiuc. Thanks for the suggestion